support/agent/miner_agent.go

```diff
@@ -415,6 +407,70 @@ func (ma *MinerAgent) submitPoStForDeadline(v SimState, dlIdx uint64) ([]message
 	}}, nil
 }
 
+// create a deal proposal message and notify provider of deal
+func (ma *MinerAgent) publishStorageDeals() []message {
+	if len(ma.pendingDeals) == 0 {
+		return []message{}
+	}
+
+	params := market.PublishStorageDealsParams{
+		Deals: ma.pendingDeals,
+	}
+	ma.pendingDeals = nil
+
+	return []message{{
+		From:   ma.Worker,
+		To:     builtin.StorageMarketActorAddr,
+		Value:  big.Zero(),
+		Method: builtin.MethodsMarket.PublishStorageDeals,
+		Params: &params,
+		ReturnHandler: func(_ SimState, _ message, ret cbor.Marshaler) error {
+			// add returned deal ids to be included within sectors
+			publishReturn, ok := ret.(*market.PublishStorageDealsReturn)
+			if !ok {
+				return errors.Errorf("create miner return has wrong type: %v", ret)
+			}
+
+			for idx, dealId := range publishReturn.IDs {
+				ma.dealsPendingInclusion = append(ma.dealsPendingInclusion, pendingDeal{
+					id:   dealId,
+					size: params.Deals[idx].Proposal.PieceSize,
+					ends: params.Deals[idx].Proposal.EndEpoch,
+				})
+			}
+			return nil
+		},
+	}}
+}
+
+func (ma *MinerAgent) updateMarketBalance() []message {
```

Comment:
I am noting duplication with client agent and that there exists an opportunity to reduce code with a shared method though the current level of duplication does seem fine.

Reply:
I took a stab at this. It definitely can be done, but the additional interfaces and implementations will require many more lines of code that simply duplicating the methods and/or make configuration more complex. I'm going to hold off on this till I figure out a better way, or we want to add something that would cause more duplication.